rules/manager.go

```diff
@@ -457,67 +457,46 @@ func (c ruleDependencyController) AnalyseRules(rules []Rule) {
 // Its purpose is to bound the amount of concurrency in rule evaluations to avoid overwhelming the Prometheus
 // server with additional query load. Concurrency is controlled globally, not on a per-group basis.
 type RuleConcurrencyController interface {
-	// Allow determines whether any concurrent evaluation slots are available.
-	// If Allow() returns true, then Done() must be called to release the acquired slot.
-	Allow() bool
+	// Allow determines if the given rule is allowed to be evaluated concurrently.
+	// If Allow() returns true, then Done() must be called to release the acquired slot and corresponding cleanup is done.
+	Allow(ctx context.Context, group *Group, rule Rule) bool
```

Comment:
I would comment that `Group` and `Rule` should not be retained by the actual controller implementation. I think it's risky retaining them.

Reply:
Done in https://github.com/prometheus/prometheus/pull/14491/commits/5229afd8307c77a4930156ea44dcf2e1f160f019